File: src/ios/ConfigManager.m
//
//  ConfigManager.m
//  emission
//
//  Created by Kalyanaraman Shankari on 3/25/16.
//
//

#import "ConfigManager.h"
#import "BEMBuiltinUserCache.h"
#import "ConsentConfig.h"
#import "LocalNotificationManager.h"

#define SENSOR_CONFIG_KEY @"key.usercache.sensor_config"
#define CONSENT_CONFIG_KEY @"key.usercache.consent_config"

@implementation ConfigManager

static LocationTrackingConfig *cachedConfig;
static ConsentConfig *cachedConsentConfig;
static NSDictionary *cachedDeploymentConfig;

// corresponds to getConfig in ConfigManager.java
+ (LocationTrackingConfig*) instance {
    if (cachedConfig == NULL) {
        cachedConfig = [self readFromCache];
        if (cachedConfig == NULL) {
            // This is still NULL, which means that there is no document in the usercache.
            // Let us add a dummy one based on the default settings
            // we don't want to save it to the database because then it will look like a user override
            cachedConfig = [self getConfigDefault];
        }
    }
    return cachedConfig;
}

+ (LocationTrackingConfig*) readFromCache {
    @try {
        LocationTrackingConfig* cachedConfig = (LocationTrackingConfig*)[[BuiltinUserCache database] getDocument:SENSOR_CONFIG_KEY wrapperClass:[LocationTrackingConfig class]];
        [LocalNotificationManager addNotification:[NSString stringWithFormat:@"in readFromCache, cached tracking config = %@", cachedConfig]];
        return cachedConfig;
    } @catch (NSException *exception) {
        [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Exception while reading sensor config, returning NULL: %@", exception]];
        return NULL;
    }
}

+ (void) updateConfig:(LocationTrackingConfig*) newConfig {
    [[BuiltinUserCache database] putReadWriteDocument:SENSOR_CONFIG_KEY value:newConfig];
    cachedConfig = newConfig;
}

+ (NSDictionary*) getDeploymentConfig {
    if (cachedDeploymentConfig == NULL) {
        @try {
            cachedDeploymentConfig = [[BuiltinUserCache database] getDocument:@"config/app_ui_config" withMetadata:false];
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"in getDeploymentConfig, deployment config = %@", cachedDeploymentConfig]];
        } @catch (NSException *exception) {
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Exception while reading deployment config, returning NULL: %@", exception]];
        }
    }
    return cachedDeploymentConfig;
}

+ (LocationTrackingConfig*) getConfigDefault {
    NSDictionary* deploymentConfig = [self getDeploymentConfig];
    if (deploymentConfig != NULL && [[deploymentConfig allKeys] containsObject:@"tracking"]) {
        @try {
            NSDictionary* tracking = [deploymentConfig valueForKey:@"tracking"];
            LocationTrackingConfig* cfg = [LocationTrackingConfig new];
            [cfg setValuesForKeysWithDictionary:tracking];
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Created default tracking config with deployment-specific values"]];
            return cfg;
        } @catch (NSException *exception) {
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"Exception while reading deployment tracking config; will use built-in default: %@", exception]];
        }
    }
    return [LocationTrackingConfig new];
}

+ (ConsentConfig*) getConsentConfig {
    if (cachedConsentConfig == NULL) {
        // TODO: should we put this in a try/catch block like the instance singleton
        cachedConsentConfig = (ConsentConfig*)[[BuiltinUserCache database] getDocument:CONSENT_CONFIG_KEY wrapperClass:[ConsentConfig class]];
    }
    return cachedConsentConfig;
}

+ (ConsentConfig*) getPriorConsent {
    ConsentConfig* currConfig = (ConsentConfig*)[[BuiltinUserCache database] getDocument:CONSENT_CONFIG_KEY wrapperClass:[ConsentConfig class]];
    [LocalNotificationManager addNotification:[NSString stringWithFormat:@"in getPriorConsent, currConfig = %@", currConfig]];
    return currConfig;
}

+ (BOOL) isConsented:(NSString*)reqConsent {
    @try {
        ConsentConfig* currConfig = [self getConsentConfig];
            [LocalNotificationManager addNotification:[NSString stringWithFormat:@"checking consent with reqConsent = %@, cached currConfig.approval_date = %@",
                reqConsent, currConfig.approval_date]];

        if ([reqConsent isEqualToString:currConfig.approval_date]) {
                [LocalNotificationManager addNotification:@"isConsented = YES"];
            return YES;
        } else {
                [LocalNotificationManager addNotification:@"isConsented = NO"];
            return NO;
        }
    } @catch (NSException *exception) {
        [LocalNotificationManager addNotification:@"isConsented = exception"];
        return false;
    }
}

+ (void) setConsented:(ConsentConfig*)newConsent {
    [[BuiltinUserCache database] putReadWriteDocument:CONSENT_CONFIG_KEY value:newConsent];
}

@end

File: src/android/ConfigManager.java
package edu.berkeley.eecs.emission.cordova.tracker;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSyntaxException;

import org.apache.cordova.ConfigXmlParser;

import java.util.HashMap;

import edu.berkeley.eecs.emission.R;
import org.json.JSONException;
import org.json.JSONObject;

import edu.berkeley.eecs.emission.cordova.tracker.wrapper.ConsentConfig;
import edu.berkeley.eecs.emission.cordova.tracker.wrapper.LocationTrackingConfig;
import edu.berkeley.eecs.emission.cordova.unifiedlogger.Log;
import edu.berkeley.eecs.emission.cordova.unifiedlogger.NotificationHelper;
import edu.berkeley.eecs.emission.cordova.usercache.UserCacheFactory;

/**
 * Created by shankari on 3/25/16.
 */
public class ConfigManager {
    private static String TAG = "ConfigManager";
    private static LocationTrackingConfig cachedConfig;
    private static ConsentConfig cachedConsent;
    private static JSONObject cachedDeploymentConfig;

    public static LocationTrackingConfig getConfig(Context context) {
        if (cachedConfig == null) {
            cachedConfig = readFromCache(context);
            if (cachedConfig == null) {
                // Still null, so there was no valid config in the usercache
                // Return the default config (with deployment-specific overrides if any)
                // Note: we do not store the default in the usercache, as it is not user-given
                cachedConfig = getConfigDefault(context);
            }
        }
        return cachedConfig;
    }

    private static LocationTrackingConfig readFromCache(Context context) {
        try {
            LocationTrackingConfig cfg = UserCacheFactory.getUserCache(context).getDocument(R.string.key_usercache_sensor_config, LocationTrackingConfig.class);
            Log.i(context, TAG, "in readFromCache, cached tracking config = " + cfg);
            return cfg;
        } catch (JsonParseException e) {
            Log.e(context, TAG, "Exception while reading sensor config json, returning null: " + e);
            return null;
        }
    }

    protected static void updateConfig(Context context, LocationTrackingConfig newConfig) {
        UserCacheFactory.getUserCache(context).putReadWriteDocument(R.string.key_usercache_sensor_config, newConfig);
        cachedConfig = newConfig;
    }

    public static JSONObject getDeploymentConfig(Context context) {
        if (cachedDeploymentConfig == null) {
            try {
                cachedDeploymentConfig = (JSONObject) UserCacheFactory.getUserCache(context).getDocument("config/app_ui_config", false);
                Log.i(context, TAG, "In getDeploymentConfig, deployment config = " + cachedDeploymentConfig);
            } catch (JSONException e) {
                Log.e(context, TAG, "Found error " + e + "parsing deployment config json, returning null");
            }
        }
        return cachedDeploymentConfig;
    }

    public static LocationTrackingConfig getConfigDefault(Context context) {
        JSONObject deploymentConfig = getDeploymentConfig(context);
        if (deploymentConfig != null && deploymentConfig.has("tracking")) {
            try {
                JSONObject tracking = deploymentConfig.getJSONObject("tracking");
                LocationTrackingConfig cfg = new Gson().fromJson(tracking.toString(), LocationTrackingConfig.class);
                Log.d(context, TAG, "Created default tracking config with deployment-specific values");
                return cfg;
            } catch (JSONException|JsonParseException e) {
                Log.e(context, TAG, "Exception while parsing tracking config from deployment config, will return built-in default config: " + e);
            }
        }
        return new LocationTrackingConfig();
    }

    public static String getReqConsent(Context ctxt) {
        ConfigXmlParser parser = new ConfigXmlParser();
        parser.parse(ctxt);
        String reqConsent = parser.getPreferences().getString("emSensorDataCollectionProtocolApprovalDate", null);
        return reqConsent;
    }

    public static ConsentConfig getConsent(Context context) throws JsonParseException {
        if (cachedConsent == null) {
            cachedConsent = UserCacheFactory.getUserCache(context)
                .getDocument(R.string.key_usercache_consent_config, ConsentConfig.class);
            Log.i(context, TAG, "Read cached consent "+cachedConsent+" from database");
        }
        return cachedConsent;
    }

    public static boolean isConsented(Context context, String reqConsent) {
        try {
            ConsentConfig currConfig = getConsent(context);
            return currConfig != null && reqConsent.equals(currConfig.getApproval_date());
        } catch(JsonParseException e) {
            Log.e(context, TAG, "Found error " + e + "parsing consent json, re-prompting user");
            return false;
        }
    }

    public static void setConsented(Context context, ConsentConfig newConsent) {
        UserCacheFactory.getUserCache(context)
                .putReadWriteDocument(R.string.key_usercache_consent_config, newConsent);
    }
}
